### DISCUSSION TITLE
AOT, Dictionary Structure and Stack Optimization.

### DISCUSSION THREAD

**Discussion by jemo07** [ForthHub/discussion#151] [Ideas]
Given the rich heritage of the Forth language and its unique stack-based design, I've been pondering some intriguing possibilities for Ahead-Of-Time (AOT) compilation. How feasible would it be to blend the concepts of a linearized dictionary with procedural mapping, and a more optimized stack operation system that employs joint procedures?

For instance, let's say our traditional Forth dictionary exists in a linearized structure, where words are stored in a flat array. When word A references word B, it does so not by name, but by a direct offset or index. This would result in direct function or procedure calls in the target assembly rather than indirect dictionary lookups, paving the way for significant runtime performance enhancements.

Moreover, when considering stack operations, rather than translating every stack command one-by-one, what if the interpreter looks ahead to identify a sequence of operations? This lookahead would continue until it encounters a literal or another stop criterion. The recognized sequence, like 'dup swap' or 'dup drop', would then be hashed into an index of common joint procedures. When this hash matches a sequence in our linearized dictionary, it directly points to an optimized translation of that sequence. This 'joint procedure' approach could minimize the instructions generated and boost execution speed.

To bring this to life, envision a sequence like 'dup swap'. Instead of generating a series of machine-level instructions for 'dup' and then 'swap', the joint procedure lookup would find an optimized instruction sequence specific for 'dup swap', resulting in a much-reduced translation. Furthermore, redundancy checks could recognize sequences like 'dup drop' as no-ops and avoid generating any machine code for them.

Incorporating these ideas, could we not only reinvent Forth's compilation for modern architectures but also lay the groundwork for an innovative approach to compiling other stack-based languages? What are the potential pitfalls and challenges to be aware of in this venture?
I wrote a gist here to further endeavor on the idea:  https://bit.ly/45uC90Z

**Comment by ruv** [comment]
It seems, some Forth systems have been using such techniques for decades already.

[Peephole optimization](https://en.wikipedia.org/wiki/Peephole_optimization) and inlining are used in many Forth system. As a more sophisticated example of optimization — see optimized code that [lxf-ntf 2017](https://www.complang.tuwien.ac.at/forth/lxf-ntf2017.zip) generates (via `see foo`), which can eliminate local variables and the use of the return stack. Another example is [Dynamic Superinstructions](https://gforth.org/manual/Dynamic-Superinstructions.html)  in Gforth.

Concerning compilation of stack-based languages in general — probably, WebAssembly is attracting most of the research and development in this area now.


> what if the interpreter looks ahead to identify a sequence of operations? 

BTW, it's not necessary to do it in one pass. You can generate an intermediate representation (IR) for a word definition, and then generate machine code from this IR.

**Reply by jemo07** [reply]
Thank you for all the references, I was not aware of the gforth implementation, I will have to look a little harder at the example at first glance i did not follow it.

I'm aware of peephole optimization, and I should have made a distinction from my discussion, whhile peephole optimization typically examines a small window of instructions, the proposed approach emphasizes a broader lookahead mechanism, especially focused on recognizing joint sequences of stack operations in a Forth environment.

What I'm calling dynamic mapping, the method leans heavily on a predefined index of common joint procedures and a hashing mechanism to map sequences to optimized translations. This adds a layer of adaptability and growth potential as new sequences can be incrementally added to the index. Might be creative syndrome ( apologies if I'm missing the point )  but I believe the proposed stack optimization I'm  envisioning is tailored for the unique intricacies and needs of Forth's stack-driven operations. It's designed to be synergistic with the linearized dictionary and procedural mapping strategies I've discussed.

Again, thank you for your input, I will have to review further the context you have provided and digest it to better understand if in fact, this is a common practice I might have overlooked. I greatly appreciate all opportunities to educate me in the inner workings of Forth. 

Ah, I just saw this is an executable, sorry I only run on OSX and Linux... will test it when possible. 

How do I dump data from the gforth dictionary to experiment further? >link seem to be missing.

**Reply by ruv** [reply]
> I just saw this is an executable, sorry I only run on OSX and Linux... will test it when possible.

lxf-ntf ZIP contains binaries for Linux and Windows. The source code isn't provided by the author.

> How do I dump data from the gforth dictionary to experiment further? `>link` seem to be missing.

The standard [word](https://forth-standard.org/standard/tools/TRAVERSE-WORDLIST) `traverse-wordlist ( i*x xt wid -- j*x )` can be used to enumerate words in a word list, or `words` to print the content of the first word list in the search order. See also  [Examining compiled code](https://gforth.org/manual/Examining-compiled-code.html) in Gforth manual.

**Reply by jemo07** [reply]
THANK YOU!!! :+1:

**Comment by ruv** [comment]
> I wrote a gist here to further endeavor on the idea: https://bit.ly/45uC90Z

At this gist your write:

> Traditional Dictionary:
>
> ```
> Word "PRINT" -> Address 0x1000
> Word "DOUBLE" -> Address 0x2000
> Word "ADD" -> Address 0x3000
> 
> At Address 0x1000:
>     PUSH address of "DOUBLE" (name or string)
>     CALL DictionaryLookup
>     CALL ReturnAddressFromLookup ; indirectly call DOUBLE
> ```


It looks like [late binding](https://en.wikipedia.org/wiki/Late_binding). It has never been _traditional_ dictionary. Forth has used early binding from the beginning, and every published standard dictates this for Forth.   Late binding was used by Chuck Moor in his experiments and precursors to Forth, see his paper [Forth - The Early Years](https://colorforth.github.io/HOPL.html).

**Reply by jemo07** [reply]
Sorry I was been lazy, I just wrote a symbolic concept, never gave it a thought, but I will clarify thank you. 

```
Word "PRINT" -> Address 0x1000
Word "DOUBLE" -> Address 0x2000
Word "ADD" -> Address 0x3000

At Address 0x1000 (PRINT):
    CALL 0x2000 ; Direct call to DOUBLE

At Address 0x2000 (DOUBLE):
    ; (The implementation of DOUBLE might involve some operations and possibly calling other words.)

At Address 0x3000 (ADD):
    ; Actual machine  code for ADD
```
I hope this is a better representation... Thank you.
